### DISCUSSION TITLE
KCL vs CUE

### DISCUSSION THREAD

**Discussion by mihaigalos** [kcl-lang/kcl#597] [Q&A] [answered]
Hi,

I was going through the [cited](https://kcl-lang.io/docs/user_docs/getting-started/intro/#vs-cue) reason as to why `kcl` is better than `cue`:

>  However, since CUE performs all constraint checks at runtime, there may be performance bottlenecks in large-scale modeling scenarios.

I am not sure that is completely true, maybe you can elaborate?

My understanding is that `cue` can perform [client-side validation](https://cuelang.org/docs/usecases/validation/#client-side-validation).

**Comment by Peefy** [comment, marked as answer]
Hi @mihaigalos . I think I understand what you mean. In fact, both KCL and CUE can perform client-side validation in scenarios. The runtime here refers to the technical concept, specifically the runtime of the language itself, such as Python and Go performing type checks at runtime and compile time, which to some extent leads to Go running faster than Python. I will revise the contents on the website in the future to avoid misunderstandings. Thank you for pointing out this content. ❤️

**Comment by ufutenina428-ops** [comment]
I see what you're getting at. You're right to question that claim. While CUE does perform runtime checks, the client-side validation isn't necessarily a huge performance bottleneck in most cases. The concern with CUE might stem from its flexibility—since it's so dynamic, it could add some overhead, especially with complex models but better just chill and log on to https://neospin.casinologin.mobi/ for slot games
